wip-0021: Add StringParseXMLMap to RADON operators (#79)
witnet/WIPs: wip-stringparsexml.md

<pre>
  WIP: WIP-00XX
  Layer: Consensus (hard fork)
  Title: Add StringParseXMLMap to RADON operators

>   Authors: Luis Rubio <luisrwitnet.foundation>

Reviewer: Malformed email

  Discussions-To: `#dev-general` channel on Witnet Community's Discord server
  Status: Draft
  Type: Standards Track
  Created: 2021-12-14
  License: BSD-2-Clause
</pre>

## Abstract

The `StringParseXMLMap` operator is proposed to be included in the operators catalog available in the RAD engine. Data requests
will then support XML parsing as part of the **retrieval** scripts.

## Motivation and rationale

In order to increase the flexibility and potential use cases of the Witnet protocol, it is paramount to continue 
expanding the catalog of available operators.

This document proposes adding support for **XML parsing**, thus allowing the protocol to extract information from XML APIs
or XML-encapsulated strings to be retrieved through data requests.

## Specification

The `StringParseXMLMap` operator converts a `String` (which is assumed to represent a valid XML document) into a `Map`, 
following the next rules:
- Every element MUST be turned into a `Map` in which the tag name will be inserted as the `_tag` key, **attributes** 
  would use the *@* symbol to precede their key name, and the elements inside would be added in a *RadonArray* with 
  the key *_children*.
  
```xml

> <tag1 attr1=001>

Reviewer 2: ```suggestion
<tag1 attr1="001">
```

    <in_tag1>[...]</in_tag1>
    <in_tag2>[...]</in_tag2>
</tag1>
```
```rust
{
    _label: "tag1"
    @attr1: "001"
    _children: [
        {in_tag1[...]}, 
        {in_tag2[...]},
    ]
}
```

- Leaf elements that have no attributes or children other than textual content MUST be added as an entry in the `Map` of
  their parent element using their tag as the keys and their textual content as the values.

```xml

> <tag1 attr1=001>

Reviewer 2: ```suggestion
<tag1 attr1="001">
```

</tag1>
```
```rust
{
    tag1: {
        @attr1: "001",
        in_tag1: "Hello",
        in_tag2: {...},
    }
}
```

> - Elements with the same tag name would be added as a `RadonArray` with the tag name in common as a *key* of the `Map`.

Reviewer 2: ```suggestion
- Elements with the same tag name MUST be collected into a `RadonArray` with the tag name in common as a *key* of the parent `Map`.
```

```xml
<tag1 attr1="001">
    <in_tag>[...]</in_tag>
    <in_tag>[...]</in_tag>
</tag1>
```
```rust
{
    tag1: {
        @attr1: "001",
        in_tag: [
            {...},

>             {...},

Reviewer 2: Please use specific examples so that it is clear how the elements are collected into the array.

> <tag1 attr1=001>

Reviewer 2: ```suggestion
<tag1 attr1="001">
```

        ],
    }
}
```

> - Elements that have several children with text content and with or without attributes MUST be included using the 
>   key `_text`.

Reviewer 2: ```suggestion
- Text children elements MUST be included into the parent `Map` using the `_text` key.
```

    _children: [
        {
            _label: "in_tag1"
            _text#0: "Hello"
            _text#1: "Witnet"
            in_in_tag1: "World"
        },
        {
            _label: "in_tag2"
            @attr: "hello"
            _text#0: "World"
        },
    ]
}
```  

- Elements with only text children or without children at all, SHOULD NOT include the `_children` key in the `Map`.

```xml

> <tag1 attr1=001>

Reviewer 2: ```suggestion
<tag1 attr1="001">
```

<pre>
  WIP: WIP-00XX
  Layer: Consensus (hard fork)
  Title: Add StringParseXMLMap to RADON operators
  Authors: Luis Rubio <luisrwitnet.foundation>
  Discussions-To: `#dev-general` channel on Witnet Community's Discord server
  Status: Draft
  Type: Standards Track
  Created: 2021-12-14
  License: BSD-2-Clause
</pre>

## Abstract

The **StringParseXMLMap** is proposed to be included in the operators catalog available in the RAD engine. Data requests
will then support XML parsing as part of the **retrieval** scripts.

## Motivation and rationale

In order to increase the flexibility and potential use cases of the Witnet protocol, it is paramount to continue 

> expanding the catalog of available .

Reviewer: of available what

This document proposes adding support for **XML parsing** allowing the protocol to extract information from XML APIs
or XML-encapsulated strings to be retrieved in Witnet by witnesses

## Specification

> The *StringParseXMLMap* operator converts a *RadonString* (where the XML is defined) in a *RadonMap* following the next

Reviewer 2: WIPs should use more generic type naming: 
```suggestion
The `StringParseXMLMap` operator converts a `String` (which is assumed to represent a valid XML document) into a `Map`, following the next
```

rules:

> - Every element will be defined as a *RadonMap* where the **tag** information would be inserted with the *_label* key, 

Reviewer 2: What do you think of actually using `_tag` instead of `_label`? It looks like that's the proper XML name. Sorry for suggesting `label` in the first place.

```suggestion
- Every element MUST be turned into a `Map` in which the tag name will be inserted as the `_tag` key, 
```

  **attributes** would use the *@* symbol to precede their key name, and the elements inside would be added in a 
  *RadonArray* with the key *_children*.
  
```xml
<tag1 attr1=001>
    <in_tag1>[...]</in_tag1>
    <in_tag2>[...]</in_tag2>
</tag1>
```
```rust
{
    _label: "tag1"
    @attr1: "001"
    _children: [
        {in_tag1[...]}, 
        {in_tag2[...]},
    ]
}
```

> - Elements with text content that have no attributes or more elements inside (aka text children) would be added as a 
>   part of the map, using the *tag* as a key and the *content* as a value

Reviewer 2: ```suggestion
- Leaf elements that have no attributes or children other than textual content MUST be added as an entry in the `Map` of their parent element using their tag as the keys and their textual content as the values.
```

```xml
<tag1 attr1=001>
    <in_tag1>[...]</in_tag1>
    <in_tag2>[...]</in_tag2>
    <in_tag3>Hello</in_tag3>
</tag1>
```
```rust
{
    _label: "tag1"
    @attr1: "001"
    in_tag3: "Hello"
    _children: [
        {in_tag1[...]}, 
        {in_tag2[...]},
    ]
}
```

> - In the case of elements that have several elements, some of them with text content, or text content with attributes.
>   Those text elements would be included directly in the map with the key *_test#n* where *n* is the text counter

Reviewer 2: ```suggestion
- Elements that have several children with text content and with or without attributes MUST include their children directly into the *Map* that represents them using the key `_text#n` where `n` is the indexed position of each of the children elements.
```

Reviewer 2: Now I wonder... what's the RADON representation of this perfectly valid XML document? :thinking: 
```xml
<tag1 attr1=001>
    <in_tag1>
        Hello
        <in_in_tag1>World</in_in_tag1>
        Witnet
        <in_in_tag1>GM</in_in_tag1>
        frens
    </in_tag1>
    <in_tag2 attr="hello">World</in_tag2>
</tag1>
```

Reviewer: Error because of missing `xmlns`. And error because `attr1=001` is invalid, must be `attr1="001"`, at least according to the library we are using. But ignoring that, this is the output:

```
RadonMap {
    value: {
        "@attr1": String(RadonString {
            value: "001"
        }),
        "_children": Array(RadonArray {
            value: [Map(RadonMap {
                value: {
                    "_label": String(RadonString {
                        value: "in_tag1"
                    }),
                    "_text#0": String(RadonString {
                        value: "Hello"
                    }),
                    "_text#1": String(RadonString {
                        value: "Witnet"
                    }),
                    "_text#2": String(RadonString {
                        value: "frens"
                    }),
                    "in_in_tag1": String(RadonString {
                        value: "GM"
                    })
                }
            }), Map(RadonMap {
                value: {
                    "@attr": String(RadonString {
                        value: "hello"
                    }),
                    "_label": String(RadonString {
                        value: "in_tag2"
                    }),
                    "_text#0": String(RadonString {
                        value: "World"
                    })
                }
            })],
            is_homogeneous: true
        }),
        "_label": String(RadonString {
            value: "tag1"
        }),
        "_xmlns": String(RadonString {
            value: "https://witnet.io/"
        })
    }
}
```

Author: O_O, we lost the first in_in_tag1

```xml
<tag1 attr1=001>
    <in_tag1>
        Hello
        <in_in_tag1>World</in_in_tag1>
        Witnet
    </in_tag1>
    <in_tag2 attr="hello">World</in_tag2>
</tag1>
```
```rust
{
    _label: "tag1"
    @attr1: "001"
    _children: [
        {
            _label: "in_tag1"
            _text#0: "Hello"
            _text#1: "Witnet"
            in_in_tag1: "World"
        },
        {
            _label: "in_tag2"
            @attr: "hello"
            _text#0: "World"
        },
    ]
}
```  

> - Elements with only text children or without elements inside, would not have the key *_children*.

Reviewer 2: ```suggestion
- Elements with only text children or without children at all, SHOULD NOT include the `_children` key in the `Map`.
```

As a result:

- Non-implementers MAY NOT get their blocks and transactions accepted by implementers.
- Implementers MAY NOT get their valid blocks accepted by non-implementers.
- Non-implementers MAY consolidate different blocks and superblocks than implementers.

Due to this last point, this MUST be considered a consensus-critical protocol improvement. An adoption plan MUST be proposed, setting an activation date that gives miners enough time in advance to upgrade their existing clients.

### Libraries, clients, and interfaces

The protocol improvements proposed herein will affect any library or client implementing any functionality related to the RAD engine. For example, this is the case of the [Sheikah][sheikah] data request editor and the `witnet-requests-js` Javascript library.

## Reference Implementation

A reference implementation for the proposed protocol improvement can be found in the [pull request #2138](https://github.com/witnet/witnet-rust/pull/2138) of the [witnet-rust] repository.

## Adoption Plan

The activation date is proposed for January 18 2022 at 9 am UTC, that is, protocol epoch #885040.

> From the perspective of TAPI, the signaling bit is `2` (`0b00000000000000000000000000000100`) and the `first_signaling_epoch` is set to #885040.

Reviewer 2: This would be better removed for the time being.

<pre>
  WIP: WIP-00XX
  Layer: Consensus (hard fork)
  Title: Add StringParseXMLMap to RADON operators
  Authors: Luis Rubio <luisrwitnet.foundation>
  Discussions-To: `#dev-general` channel on Witnet Community's Discord server
  Status: Draft
  Type: Standards Track
  Created: 2021-12-14
  License: BSD-2-Clause
</pre>

## Abstract

The **StringParseXMLMap** is proposed to be included in the operators catalog available in the RAD engine. Data requests
will then support XML parsing as part of the **retrieval** scripts.

## Motivation and rationale

In order to increase the flexibility and potential use cases of the Witnet protocol, it is paramount to continue 
expanding the catalog of available .

> This document proposes to add the support of **XML parsing** allowing to the protocol extract information from XML APIs

Reviewer 3: ```suggestion
This document proposes adding support for **XML parsing** allowing the protocol to extract information from XML APIs or XML-encapsulated strings to be retrieved in Witnet by witnesses.
```

or XML-encapsulated strings to be retrieved in Witnet by witnesses

## Specification

The *StringParseXMLMap* operator converts a *RadonString* (where the XML is defined) in a *RadonMap* following the next
rules:
- Every element will be defined as a *RadonMap* where the **tag** information would be inserted with the *_label* key, 
  **attributes** would use the *@* symbol to precede their key name, and the elements inside would be added in a 
  *RadonArray* with the key *_children*.
  
```xml
<tag1 attr1=001>
    <in_tag1>[...]</in_tag1>
    <in_tag2>[...]</in_tag2>
</tag1>
```
```rust
{
    _label: "tag1"
    @attr1: "001"
    _children: [
        {in_tag1[...]}, 
        {in_tag2[...]},
    ]
}
```

> - Elements with text content that have not attributes or more elements inside (aka text children) would be added as a 

Reviewer 3: ```suggestion
- Elements with text content that have no attributes or more elements inside (aka text children) would be added as a 
```

  part of the map, using the *tag* as a key and the *content* as a value

```xml
<tag1 attr1=001>
    <in_tag1>[...]</in_tag1>
    <in_tag2>[...]</in_tag2>
    <in_tag3>Hello</in_tag3>
</tag1>
```
```rust
{
    _label: "tag1"
    @attr1: "001"
    in_tag3: "Hello"
    _children: [
        {in_tag1[...]}, 
        {in_tag2[...]},
    ]
}
```

> - In case of elements that have several elements, some of them with text content, or text content with attributes.

Reviewer 3: ```suggestion
- In the case of elements that have several elements, some of them with text content, or text content with attributes.
```

  Those text elements would be included directly in the map with the key *_test#n* where *n* is the text counter

```xml
<tag1 attr1=001>
    <in_tag1>
        Hello
        <in_in_tag1>World</in_in_tag1>
        Witnet
    </in_tag1>
    <in_tag2 attr="hello">World</in_tag2>
</tag1>
```
```rust
{
    _label: "tag1"
    @attr1: "001"
    _children: [
        {
            _label: "in_tag1"
            _text#0: "Hello"
            _text#1: "Witnet"
            in_in_tag1: "World"
        },
        {
            _label: "in_tag2"
            @attr: "hello"
            _text#0: "World"
        },
    ]
}
```  

> - Elements with only text children or not any elements inside at all, would not have the key *_children*

Reviewer 3: ```suggestion
- Elements with only text children or without elements inside would not have the key *_children*.
```

```xml
<tag1 attr1=001>
    <in_tag1>Hello</in_tag1>
    <in_tag2>World</in_tag2>
</tag1>
```
```rust
{
    _label: "tag1"
    @attr1: "001"
    in_tag1: "Hello"
    in_tag2: "World"
}
```

## Backwards compatibility

- Implementer: a client that implements or adopts the protocol improvements proposed in this document.
- Non-implementer: a client that fails to or refuses to implement the protocol improvements proposed in this document.

#### Consensus cheat sheet

Upon entry into force of the proposed improvements:

> - Blocks and transactions that were considered valid by former versions of the protocol MUST continue to be considered valid.

Reviewer 3: ```suggestion
- Blocks and transactions considered valid by former versions of the protocol MUST continue to be considered valid.
```

- Blocks and transactions produced by non-implementers MUST be validated by implementers using the same rules as with those coming from implementers, that is, the new rules.
- Implementers MUST apply the proposed logic when evaluating transactions or computing their own data request eligibility.

As a result:

- Non-implementers MAY NOT get their blocks and transactions accepted by implementers.
- Implementers MAY NOT get their valid blocks accepted by non-implementers.
- Non-implementers MAY consolidate different blocks and superblocks than implementers.

Due to this last point, this MUST be considered a consensus-critical protocol improvement. An adoption plan MUST be proposed, setting an activation date that gives miners enough time in advance to upgrade their existing clients.

### Libraries, clients, and interfaces

> The protocol improvements proposed herein will affect any library or client implementing the any functionality related to the RAD engine. For example, this is the case of the [Sheikah][sheikah] data request editor and `witnet-requests-js` Javascript library.

Reviewer 3: ```suggestion
The protocol improvements proposed herein will affect any library or client implementing any functionality related to the RAD engine. For example, this is the case of the [Sheikah][sheikah] data request editor and the `witnet-requests-js` Javascript library.
```

## Reference Implementation

A reference implementation for the proposed protocol improvement can be found in the [pull request #2138](https://github.com/witnet/witnet-rust/pull/2138) of the [witnet-rust] repository.

## Adoption Plan

> An activation date is proposed for January 18 2022 at 9am UTC, that is, protocol epoch #885040.

Reviewer 3: ```suggestion
The activation date is proposed for January 18 2022 at 9 am UTC, that is, protocol epoch #885040.
```
